libfwupdplugin/fu-ifd-firmware.c

```diff
@@ -0,0 +1,434 @@
+/*
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN				"FuFirmware"
+
+#include "config.h"
+
+#include "fu-common.h"
+#include "fu-ifd-firmware.h"
+
+/**
+ * SECTION:fu-ifd-firmware
+ * @short_description: IFD firmware image
+ *
+ * An object that represents a Intel Flash Descriptor.
+ *
+ * Still left to do:
+ *
+ * - Obtain this image as part of the HSI security check
+ * - Check 0x42 is in the invalid JEDEC instruction list
+ * - Add fu_ifd_image_is_writable() if another section can write the image
+ *  - Fail HSI if the descriptor is writable!
+ *
+ * See also: #FuFirmware
+ */
+
+typedef struct {
+	guint32			 flmap0;
+	guint32			 flmap1;
+	guint32			 flmap2;
+	guint8			 nr;
+	guint8			 nc;
+	guint32			 frba;
+	guint32			 fcba;
+	guint32			 fmba;
+	guint32			 fisba;
+	guint32			 fmsba;
+	guint32			 flcomp;
+	guint32			 flill;
+	guint32			 flill1;
+	guint32			*flregs;
```

Comment:
these mean nothing to me, any chance they can be expanded to better names?  `nr` = `num_regions` or similar?

Reply:
We could; I just used the names from the spec. I'll expand them out a little.

Reply:
Oh it's from the spec?  OK..  Maybe just in the `to_string` is fine then to expand what they mean

Reply:
Nah, you're completely correct; I'm not going to know what `fisba` is in 6 months time, but i will know what `flash_ich_strap_base_addr` is

Reply:
FuIfdFirmware:
	DescriptorMap0:         0x40003
	DescriptorMap1:         0x58100208
	DescriptorMap2:         0x310330
	NumRegions:             10
	NumComponents:          1
	FlashRegionBaseAddr:    0x40
	FlashComponentBaseAddr: 0x30
	FlashMasterBaseAddr:    0x80
	FlashIchStrapBaseAddr:  0x100
	FlashMchStrapBaseAddr:  0x300
	ComponentsRcd:          0x325c00f5
	IllegalJedec:           0xad604221
	IllegalJedec1:          0xc7c4b9b7
	FlashDescriptorReg0:    0x0
	FlashDescriptorReg1:    0xfff0700
	FlashDescriptorReg2:    0x6ff0003
	FlashDescriptorReg3:    0x20001
	FlashDescriptorReg4:    0x7fff
	FlashDescriptorReg5:    0x7fff
	FlashDescriptorReg6:    0x7fff
	FlashDescriptorReg7:    0x7fff
	FlashDescriptorReg8:    0x7fff
	FlashDescriptorReg9:    0x7fff
	  FuFirmwareImage:
	  ID:                   desc
	  Data:                 0xfff
	  FuFirmwareImage:
	  ID:                   bios
	  Index:                0x1
	  Address:              0x700000
	  Data:                 0x8fffff
	  FuFirmwareImage:
	  ID:                   me
	  Index:                0x2
	  Address:              0x3000
	  Data:                 0x6fcfff
	  FuFirmwareImage:
	  ID:                   gbe
	  Index:                0x3
	  Address:              0x1000
	  Data:                 0x1fff